DVSNI Revision

jdkasten opened the pull request:
I believe this fixes issues #19 #33 #107 and #174.

It also simplifies the process and removes the call to SHA-256.

bifurcation commented:
FWIW, I think we've been doing rebases rather than merges from master.

martinthomson commented:
I think that I'd like to see this match the general form of the other challenges that @bifurcation has proposed in #169 and #170.

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
 The Domain Validation with Server Name Indication (DVSNI) validation method aims to ensure that the ACME client has administrative access to the web server at the domain name being validated, and possession of the private key being authorized.  The ACME server verifies that the operator can reconfigure the web server by having the client create a new self-signed challenge certificate and respond  ...
 
-The challenge proceeds as follows: The ACME server sends the client a random value R and a nonce used to identify the transaction.  The client responds with another random value S.  The server initiates a TLS connection on port 443 to one or more of the IPv4 or IPv6 hosts with the domain name being validated.  In the handshake, the ACME server sets the Server Name Indication extension set to "\<n ...
+The challenge proceeds as follows: The ACME server sends the client a random value R, a nonce used to identify the transaction, and an array of acceptable ports for which the challenge may be performed.  The client responds with the desired port and key used within the test certificate.  The server initiates a TLS connection on the client's chosen port to one or more of the IPv4 or IPv6 hosts wit ...
```
"and key" -> "and the key". Also, it's not clear how I deliver the key. As...?

ekr replied:
"one or more"? That's confusing.

Shouldn't this list of things the client supplies indicate the address? Why not indicate the FQDN instead?

ekr replied:
"sets the server Name Indication extension set"... superfluous set

How does the certificate contain the domain name? Can you point to the security explanation for requiring the cert to contain the domain name?

jdkasten replied:
Thanks, I think this will be more clear after I restructure the challenge towards the form of #170 as suggested by @martinthomson. 

The server's certificate contains the "token domain name" (\<R\>.acme.invalid).  It does not need to contain the domain name intended for authorization.  The "token domain name" is meant to directly tie the individual ACME challenge to the certificate.

ekr replied:
No, I mean where does it contain it. In which field?

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
 
 r (required, string):
-: A random 32-byte octet, Base64-encoded
+: A random 32-byte octet, hex-encoded
```
32-byte octet?

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
+: A random 32-byte octet, hex-encoded
 
 nonce (required, string):
 : A random 16-byte octet string, hex-encoded (so that it can be used as a DNS label)
```
Why is there a separate R and nonce and why are they different lenghts.

jdkasten replied:
They were originally separated to make the configuration of the server extremely explicit.  Making the client configure a server to respond to \<nonce\>.acme.invalid and return a certificate "only" containing the name \<R\>.acme.invalid seemed to me to be a greater sense of control.  It is essentially a misconfigured server, and requires slightly greater control over the configuration.

I agree that they should be the same length and 16-bytes should be long enough.

Note: The certificate only containing the \<R\>.acme.invalid domain was lost on one of the iterations.

ekr replied:
I don't find this argument very convincing. It seems like extra complexity that we then need to analyze.

If you want to have the server be "misconfigured" (this seems of dubious value) then why not simple require that the cert contain "<R>.cert.example.invalid" and the SNI be "<R>-sni.example.invalid"

jdkasten replied:
The original goal was to demonstrate *arbitrary* control over the server configuration. A misconfiguration seemed like the best way to guarantee that sysadmins wouldn't be duped into mistakenly provisioning the service.  Completely relying on shared hosting environments not to provision server names with "acme.invalid" seemed risky.

A simple transformation would suffice in this situation and my working version already contains it.  I am afraid I got trapped into trying to make the smallest number of protocol changes possible, instead of rethinking the challenge from scratch.  Thank you for the discussion @ekr.

jdkasten replied:
Assuming there isn't any further opposition, I will remove the SNI/cert complexity altogether.  After thinking about it for awhile longer, it will simplify the deployment in particular [web servers](http://doc.lighttpd.net/lighttpd2/mod_gnutls.html) and it does simplify the threat model.

Thanks

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
-The client responds to this Challenge by configuring a TLS server on port 443 of a server with the domain name being validated:
+The client responds to this Challenge by configuring a TLS server on the chosen port of a server with the domain name being validated:
+
+1. Generate a self-signed certificate with a subjectAltName extension containing the dNSName value "\<R\>[0:32].\<R\>[32:64].acme.invalid"
```
This seems inconvenient. Why isn't 16 bytes enough?

ekr replied:
If this is self-signed, what are the contents of subject and issuer?

ekr replied:
It seems like you should prescribe the entire contents of the cert,

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
-s (required, string):
-: A random 32-byte secret octet string, Base64-encoded
+port (required, int):
+: Desired port to perform the challenge
```
Why are we allowing ports != 443? That seems fairly hard to analyze from a security perspective. I notice that you don't even require that port < 1024.

jdkasten replied:
I was attempting to separate current CA policy from the ACME protocol.  I had intended to include a note that CAs should require port < 1024.

Let's Encrypt wouldn't be allowing anything other than 443.  The CABforum's Domain Validation proposal currently hardcodes 443.  So conforming CAs would not be allowed to verify on other ports anyway.  If you think it is best to outlaw all other ports in ACME itself, I will revert the challenge back to strictly using 443.

ekr replied:
Do you have an argument for why other ports should be allowed?

jdkasten replied:
*PROs*
Since the challenge is performed strictly within TLS, the challenge is capable of being used to verify services other than HTTP/HTTPS.  Mail services come to mind.
Allowing other ports makes the challenge more flexible without complicating Boulder code (```ports: [443]``` would be hardcoded)
Partially addresses #19 #33 and [various discussions found on the IETF mailing list] (https://mailarchive.ietf.org/arch/msg/acme/LgXR7qVLGgLqvE1Y5wHuOGhzO-g)

*CONs*
Adds an unnecessary complication for launch that will not be immediately used. (Barred by CABforum)
The addition doesn't solve everyone's problem.

Given the timing, it is probably best to keep it simple and focused.  I will remove the ports specification.

ekr commented on draft-barnes-acme.md:
```diff
@@ -1212,66 +1212,74 @@ If all of the above verifications succeed, then the validation is successful.  I
-  * It contains the domain name being validated as a subjectAltName
-  * It contains a subjectAltName matching the hex-encoding of Z, with the suffix ".acme.invalid"
+  * The public key is the same key found in the client's DVSNI response.
+  * It contains a subjectAltName of the form "\<R\>[0:32].\<R\>[32:64].acme.invalid".acme.invalid".
```
Wouldn't this all be easier if the client supplied a certificate fingerprint?

jdkasten replied:
Agreed.

jdkasten commented:
Thanks for the review and comments @ekr @bifurcation and @martinthomson.  There were some other stale parts to the challenge and i appreciate the discussion.  I will update the PR addressing all of the comments early tomorrow morning

jdkasten commented:
Greatly simplified and I tried to follow the format used within the new version of the SimpleHTTP challenge.

jdkasten commented:
I can submit a PR that rebases off of the 80 character wrap PR #179

jdkasten commented:
Created a new branch within this repository to make it easier for others to edit.  It also squashes the commit history. #180
